http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala

```diff
@@ -0,0 +1,342 @@
+  metrics: OpenTelemetryMetrics[F],
+  logger: StructuredLogger[F],
+  clusterToolToToolDao: RuntimeContainerServiceType => ToolDAO[F, RuntimeContainerServiceType],
+  ec: ExecutionContext
+) {
+  private val parallelism = 25
+
+  /** Entry point of this class; starts the async process */
+  val process: Stream[F, Unit] =
+    (Stream.sleep[F](config.checkInterval) ++ Stream.eval(
+      retrieveMetrics
+        .handleErrorWith(e => logger.error(e)("Unexpected error occurred during metric monitoring"))
+    )).repeat
+
+  private[monitor] def retrieveMetrics: F[Unit] =
+    for {
+      now <- F.realTimeInstant
+      traceId = TraceId(s"AppMetricsMonitor_${now.toEpochMilli}")
+      implicit0(appContext: Ask[F, AppContext]) = Ask.const[F, AppContext](AppContext(traceId, now))
+      _ <- retrieveAppMetrics
+      _ <- retrieveRuntimeMetrics
+    } yield ()
+
+  /** Queries the DB for all active apps and collects metrics */
+  private[monitor] def retrieveAppMetrics(implicit ev: Ask[F, AppContext]): F[Unit] = for {
+    clusters <- KubernetesServiceDbQueries.listAppsForMetrics.transaction
+    appDbStatus = countAppsByDbStatus(clusters)
+    _ <- recordMetric(appDbStatus)
+    _ <- logger.info(s"Recorded status metrics for ${appDbStatus.size} apps")
+    appHealth <- countAppsByHealth(clusters)
+    _ <- recordMetric(appHealth)
+    _ <- logger.info(s"Recorded health metrics for ${appHealth.size} apps")
+  } yield ()
+
+  /** Queries the DB for all active runtimes and collects metrics */
+  private[monitor] def retrieveRuntimeMetrics(implicit ev: Ask[F, AppContext]): F[Unit] = for {
+    runtimeSeq <- clusterQuery.listActiveForMetrics.transaction
+    runtimes = runtimeSeq.toList
+    runtimeDbStatus = countRuntimesByDbStatus(runtimes)
+    _ <- recordMetric(runtimeDbStatus)
+    _ <- logger.info(s"Recorded status metrics for ${runtimeDbStatus.size} runtimes")
+    runtimeHealth <- countRuntimesByHealth(runtimes)
+    _ <- recordMetric(runtimeHealth)
+    _ <- logger.info(s"Recorded health metrics for ${runtimeHealth.size} runtimes")
+  } yield ()
+
+  /** Transforms DB apps to AppStatusMetric and computes counts. */
+  private[monitor] def countAppsByDbStatus(
+    allClusters: List[KubernetesCluster]
+  ): Map[AppStatusMetric, Int] = {
+    val allApps = for {
+      c <- allClusters
+      n <- c.nodepools
+      a <- n.apps
+    } yield Map(
+      AppStatusMetric(c.cloudContext.cloudProvider,
+                      a.appType,
+                      a.status,
+                      getRuntimeUI(a.labels),
+                      getAzureCloudContext(c.cloudContext)
+      ) -> 1
+    )
+
+    allApps.combineAll
+  }
+
+  /** Transforms DB runtimes to RuntimeStatusMetric and computes counts. */
+  private[monitor] def countRuntimesByDbStatus(allRuntimes: List[RuntimeMetrics]): Map[RuntimeStatusMetric, Int] = {
+    val allContainers = for {
+      r <- allRuntimes
+      c <- r.containers if c != RuntimeContainerServiceType.WelderService
```

Comment:
Ah, because I wanted to have 1 `RuntimeStatusMetric` per runtime. But each runtime has a set of image types: `([jupyter or rstudio], welder)`. For this metric I just cared about reporting whether a runtime is jupyter or rstudio.

…I guess this relies on some non-intuitive assumptions of Leo internals.

Reply:
Ahhh gotcha, that makes a lot of sense, maybe a small comment would help future devs? There is so much implicit context in the inner workings of leo that I have yet to grasp 😅